mistralrs-core/src/sampler.rs

```diff
@@ -208,11 +201,18 @@ impl Sampler {
         top_k: i64,
         top_p: f32,
         return_logprobs: bool,
+        penalty: Option<Vec<f32>>,
     ) -> Result<Logprobs> {
         let mut argsort_indices = (0..probs.len()).collect::<Vec<_>>();
 
+        if let Some(penalty) = penalty {
+            for (token_id, logit) in probs.iter_mut().enumerate() {
+                *logit += penalty[token_id];
+            }
+        }
+
         // Sort by descending probability.
-        argsort_indices.sort_by(|&i, &j| probs[j].partial_cmp(&probs[i]).unwrap());
+        argsort_indices.sort_unstable_by(|&i, &j| probs[j].partial_cmp(&probs[i]).unwrap());
```

Comment:
`This sort is unstable (i.e., may reorder equal elements), in-place (i.e., does not allocate), and O(n * log(n)) worst-case.`

The benchmarks here show the difference between stable and unstable. (I don't think adding the lib is required)

https://github.com/orlp/glidesort
![image](https://github.com/EricLBuehler/mistral.rs/assets/12750442/37af3755-8576-43d3-ae6e-ae804fba0f9b)

This saves 500ms per token